pupil_src/shared_modules/annotations.py

```diff
@@ -10,220 +10,297 @@
+        if annotation_definitions is None:
+            annotation_definitions = [["My annotation", "E"]]
+        self._initial_annotation_definitions = annotation_definitions
+        self._definition_to_buttons = {}
 
-        self.new_annotation_name = "new annotation name"
-        self.new_annotation_hotkey = "e"
+        self._new_annotation_label = "new annotation label"
+        self._new_annotation_hotkey = "E"
 
-        self.current_frame = -1
+    def get_init_dict(self):
+        annotation_definitions = list(self._definition_to_buttons.keys())
+        return {"annotation_definitions": annotation_definitions}
 
     def init_ui(self):
         self.add_menu()
-        self.menu.label = "Annotations"
-        self.menu.append(ui.Text_Input("new_annotation_name", self))
-        self.menu.append(ui.Text_Input("new_annotation_hotkey", self))
-        self.menu.append(ui.Button("add annotation type", self.append_annotation))
-        self.sub_menu = ui.Growing_Menu("Events - click to remove")
-        self.menu.append(self.sub_menu)
-        self.update_buttons()
-
-    def update_buttons(self):
-        for b in self.buttons:
-            self.g_pool.quickbar.remove(b)
-            self.sub_menu.elements[:] = []
-        self.buttons = []
-
-        for e_name, hotkey in self.annotation_definitions:
-
-            def make_fire(e_name, hotkey):
-                return lambda _: self.fire_annotation(e_name)
-
-            def make_remove(e_name, hotkey):
-                return lambda: self.remove_annotation((e_name, hotkey))
-
-            thumb = ui.Thumb(
-                e_name,
-                setter=make_fire(e_name, hotkey),
-                getter=lambda: False,
-                label=hotkey,
-                hotkey=hotkey,
-            )
-            self.buttons.append(thumb)
-            self.g_pool.quickbar.append(thumb)
-            self.sub_menu.append(
-                ui.Button(e_name + " <" + hotkey + ">", make_remove(e_name, hotkey))
-            )
+        self.customize_menu()
+        self.menu.append(
+            ui.Text_Input("_new_annotation_label", self, label="New Label")
+        )
+        self.menu.append(
+            ui.Text_Input("_new_annotation_hotkey", self, label="New Hotkey")
+        )
+        self.menu.append(ui.Button("Add Annotation Type", self._on_add_annotation_clicked))
+        self._annotation_list_menu = ui.Growing_Menu(
+            "Annotation List - Click to Remove"
+        )
+        self.menu.append(self._annotation_list_menu)
+        self._create_initial_annotation_list()
+
+    def _create_initial_annotation_list(self):
+        for label, hotkey in self._initial_annotation_definitions:
+            self._add_annotation_definition(label, hotkey)
+
+    @abstractmethod
+    def customize_menu(self):
+        pass
 
     def deinit_ui(self):
-        self.menu.remove(self.sub_menu)
-        self.sub_menu = None
+        self._clear_button_quickbars()
```

Comment:
```suggestion
        self._clear_quickbar_buttons()
```

Reply:
uhh lovely Yoda speach, it should be `_clear_buttons_quickbar`, this order is strange but consistent with reverse notation, right?

Reply:
In my understanding the reverse notation only makes sense if some kind of topology is defined, i.e. if there were multiple `_clear_button_*` methods, this notation would be appropriate. Since there is only this function, there is no reason to prefer on order over the other. `_clear_quickbar_buttons` is however easier to read.

Reply:
Important is the plural/singular here, order is only a secondary thought.

Reply:
I wanted to be consistent with the other methods (add_button_x and create_button_x). I'm not a huge fan of reverse annotation, though, so in this case my personal preference would be create/add/clear_x_button everywhere